source/extensions/filters/http/cache/cache_insert_queue.cc

```diff
@@ -0,0 +1,184 @@
+    Http::ResponseTrailerMapImpl::copyFrom(*trailers_, trailers);
+  }
+
+  void send(InsertContext& context,
+            std::function<void(bool ready, bool end_stream, size_t sz)> on_complete) override {
+    // While zero isn't technically true for the size of trailers, it doesn't
+    // matter at this point because watermarks after the stream is complete
+    // aren't useful.
+    context.insertTrailers(*trailers_, [on_complete](bool ready) { on_complete(ready, true, 0); });
+  }
+
+private:
+  std::unique_ptr<Http::ResponseTrailerMap> trailers_;
+};
+
+CacheInsertQueue::CacheInsertQueue(Event::Dispatcher& dispatcher, InsertContextPtr insert_context,
+                                   size_t high_watermark_bytes, size_t low_watermark_bytes,
+                                   OverHighWatermarkCallback high, UnderLowWatermarkCallback low,
+                                   AbortInsertCallback abort)
+    : dispatcher_(dispatcher), insert_context_(std::move(insert_context)),
+      high_watermark_bytes_(high_watermark_bytes), low_watermark_bytes_(low_watermark_bytes),
+      over_high_watermark_callback_(high), under_low_watermark_callback_(low),
+      abort_callback_(abort) {}
+
+void CacheInsertQueue::insertHeaders(const Http::ResponseHeaderMap& response_headers,
+                                     const ResponseMetadata& metadata, bool end_stream) {
+  if (end_stream) {
+    end_in_sight_ = true;
+  }
+  // While zero isn't technically true for the size of headers, headers are
+  // typically excluded from the stream buffer limit.
+  insert_context_->insertHeaders(
+      response_headers, metadata,
+      [this, end_stream](bool ready) { onChunkComplete(ready, end_stream, 0); }, end_stream);
+  chunk_in_flight_ = true;
+}
+
+void CacheInsertQueue::insertBody(const Buffer::Instance& chunk, bool end_stream) {
+  if (end_stream) {
+    end_in_sight_ = true;
+  }
+  if (chunk_in_flight_) {
+    size_t sz = chunk.length();
+    queue_size_bytes_ += sz;
+    chunks_.push_back(std::make_unique<CacheInsertChunkBody>(chunk, end_stream));
+    if (!sent_watermark_ && queue_size_bytes_ > high_watermark_bytes_) {
+      over_high_watermark_callback_();
+      sent_watermark_ = true;
+    }
+  } else {
+    insert_context_->insertBody(
+        Buffer::OwnedImpl(chunk),
+        [this, end_stream](bool ready) { onChunkComplete(ready, end_stream, 0); }, end_stream);
+    chunk_in_flight_ = true;
+  }
+}
+
+void CacheInsertQueue::insertTrailers(const Http::ResponseTrailerMap& trailers) {
+  end_in_sight_ = true;
+  if (chunk_in_flight_) {
+    chunks_.push_back(std::make_unique<CacheInsertChunkTrailers>(trailers));
+  } else {
+    insert_context_->insertTrailers(trailers,
+                                    [this](bool ready) { onChunkComplete(ready, true, 0); });
+    chunk_in_flight_ = true;
+  }
+}
+
+void CacheInsertQueue::onChunkComplete(bool ready, bool end_stream, size_t sz) {
+  dispatcher_.post([this, ready, end_stream, sz]() {
```

Comment:
can you comment in the code on what thread this may be called from, which requires you to post? I think I can imagine what the answer is but it'd be good to confirm with a comment in the code.

Reply:
Done.